flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/FunctionDefinitionFactory.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.factories;
+
+import org.apache.flink.table.catalog.CatalogFunction;
+import org.apache.flink.table.functions.FunctionDefinition;
+import org.apache.flink.table.types.DataType;
+
+/**
+ * A factory to create {@link FunctionDefinition} based on string-based properties.
+ * See also {@link TableFactory} for more information.
+ */
+public interface FunctionDefinitionFactory extends TableFactory {
+
+	/**
+	 * Creates a {@link FunctionDefinition} from given {@link CatalogFunction}.
+	 *
+	 * @param name name of the {@link CatalogFunction}
+	 * @param catalogFunction the catalog function
+	 * @param constantArguments arguments of a function call (only literal arguments
+	 *                  are passed, nulls for non-literal ones)
+	 * @param argTypes types of arguments
+	 * @return a {@link FunctionDefinition}
+	 */
+	FunctionDefinition createFunctionDefinition(
+		String name, CatalogFunction catalogFunction, Object[] constantArguments, DataType[] argTypes);
```

Comment:
Should we change the first parameter to ObjectPath functionPath? It's possible to have functions with the same name but in different databases in a catalog.

Reply:
Why does the factory needs information about the database? Actually, I also wonder why the `TableSourceFactory` has access to `ObjectPath`. It mixes two separate layers.

Reply:
Currently HiveSource/Sink need database name and table name to access Hive metastore directly to fetch additional information needed to read/write Hive table, which is why HiveTableFactory provides ObjectPath. Note that CatalogTable itself doesn't carry over this info.